File: scripts/lib/tripo.ts
/**
 * Minimal Tripo API client.
 *
 * Tripo is task-based: you POST a task, poll it until it succeeds, then download
 * the result. Everything here is that loop plus typed helpers for the task types
 * this project uses.
 *
 * Needs TRIPO_API_KEY in .env (Bun loads it automatically).
 */
const BASE = "https://api.tripo3d.ai/v2/openapi";

const key = () => {
  const k = process.env.TRIPO_API_KEY;
  if (!k) throw new Error("TRIPO_API_KEY missing from .env");
  return k;
};

const headers = () => ({
  Authorization: `Bearer ${key()}`,
  "Content-Type": "application/json",
});

export interface TaskResult {
  task_id: string;
  status: string;
  progress: number;
  output: Record<string, unknown>;
  input?: Record<string, unknown>;
}

export async function balance(): Promise<number> {
  const r = await fetch(`${BASE}/user/balance`, { headers: headers() });
  const j = (await r.json()) as { data?: { balance: number } };
  return j.data?.balance ?? -1;
}

export async function createTask(body: Record<string, unknown>): Promise<string> {
  const r = await fetch(`${BASE}/task`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify(body),
  });
  const j = (await r.json()) as { code: number; message?: string; data?: { task_id: string } };
  if (j.code !== 0 || !j.data) {
    throw new Error(`tripo task failed (${j.code}): ${j.message ?? "unknown"}`);
  }
  return j.data.task_id;
}

export async function getTask(id: string): Promise<TaskResult> {
  const r = await fetch(`${BASE}/task/${id}`, { headers: headers() });
  const j = (await r.json()) as { code: number; message?: string; data?: TaskResult };
  if (j.code !== 0 || !j.data) throw new Error(`tripo poll failed: ${j.message}`);
  return j.data;
}

/** Poll until the task leaves the queue. Tripo tasks take 30s-3min. */
export async function waitFor(id: string, label = "task"): Promise<TaskResult> {
  const started = Date.now();
  let last = -1;
  for (;;) {
    const t = await getTask(id);
    if (t.progress !== last) {
      last = t.progress;
      const secs = ((Date.now() - started) / 1000).toFixed(0);
      process.stdout.write(`\r  ${label}: ${t.status} ${t.progress}%  (${secs}s)   `);
    }
    if (t.status === "success") {
      console.log(`\r  ${label}: done in ${((Date.now() - started) / 1000).toFixed(0)}s          `);
      return t;
    }
    if (["failed", "cancelled", "banned", "expired", "unknown"].includes(t.status)) {
      console.log("");
      throw new Error(`${label} ended as "${t.status}"`);
    }
    await new Promise((r) => setTimeout(r, 4000));
  }
}

/** Pull the first downloadable URL out of a task's output blob. */
export function outputUrl(t: TaskResult): string | null {
  const out = t.output as Record<string, unknown>;
  for (const k of ["pbr_model", "model", "rigged_model", "animated_model", "base_model"]) {
    const v = out[k];
    if (typeof v === "string" && v.startsWith("http")) return v;
    if (v && typeof v === "object" && "url" in (v as Record<string, unknown>)) {
      const u = (v as { url?: string }).url;
      if (typeof u === "string") return u;
    }
  }
  return null;
}

export async function download(url: string, dest: string) {
  const r = await fetch(url);
  if (!r.ok) throw new Error(`download failed: ${r.status}`);
  await Bun.write(dest, await r.arrayBuffer());
  return dest;
}

File: scripts/tripo.py
"""
The slice of the Tripo v2 API the asset scripts use.

Two quirks are baked in here rather than rediscovered:

  * `face_limit` is ignored by the generation tasks but honoured by
    `convert_model`, which is also the only one that decimates without wrecking
    the UVs. So every model is generated at full density and then converted.
  * `convert_model` rejects `format: "GLB"`. It wants `"GLTF"` — and hands back
    a .glb regardless.

Imported by build-heads.py and build-gear.py; not meant to be run. Every call
in here spends credits except `await_task`.
"""

from __future__ import annotations

import json
import time
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
API = "https://api.tripo3d.ai/v2/openapi"

# Tripo sits behind Cloudflare, which blocks urllib's default "Python-urllib/x.y"
# signature outright — every endpoint, including user/balance, comes back as
# HTTP 403 "error code: 1010". A normal browser User-Agent is all it wants.
UA = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"


def api_key() -> str:
    for line in (ROOT / ".env").read_text().splitlines():
        if line.startswith("TRIPO_API_KEY="):
            # .env values are sometimes quoted and sometimes not. Sending the
            # quotes along produces a plain "Authentication failed" that looks
            # exactly like an expired key, so strip them here rather than
            # debugging it a second time.
            return line.split("=", 1)[1].strip().strip("\"'")
    raise SystemExit("TRIPO_API_KEY missing from .env")


def post(path: str, payload: dict, key: str) -> dict:
    req = urllib.request.Request(
        f"{API}/{path}",
        data=json.dumps(payload).encode(),
        headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json",
                 "User-Agent": UA},
    )
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            body = json.load(r)
    except urllib.error.HTTPError as e:
        # Tripo puts the useful part in the response body, which urllib hides
        # behind a bare "HTTP Error 403: Forbidden" unless you go and read it.
        raise SystemExit(f"tripo {path} HTTP {e.code}: {e.read().decode()[:400]}") from None
    if body.get("code") != 0:
        raise SystemExit(f"tripo {path} failed: {body}")
    return body["data"]


def await_task(task_id: str, key: str, label: str) -> dict:
    while True:
        req = urllib.request.Request(
            f"{API}/task/{task_id}",
            headers={"Authorization": f"Bearer {key}", "User-Agent": UA},
        )
        with urllib.request.urlopen(req, timeout=60) as r:
            d = json.load(r)["data"]
        if d["status"] == "success":
            return d["output"]
        if d["status"] in ("failed", "cancelled", "banned", "unknown"):
            raise SystemExit(f"{label} {d['status']}: {d}")
        print(f"    {label} {d['status']} {d.get('progress', '')}%")
        time.sleep(8)


def upload_image(path: Path, key: str) -> str:
    """Uploads a PNG and returns the token that image-to-3D wants."""
    boundary = "----ceoclash"
    body = (
        f"--{boundary}\r\nContent-Disposition: form-data; name=\"file\"; "
        f"filename=\"{path.name}\"\r\nContent-Type: image/png\r\n\r\n".encode()
        + path.read_bytes()
        + f"\r\n--{boundary}--\r\n".encode()
    )
    req = urllib.request.Request(
        f"{API}/upload",
        data=body,
        headers={
            "Authorization": f"Bearer {key}",
            "Content-Type": f"multipart/form-data; boundary={boundary}",
            "User-Agent": UA,
        },
    )
    with urllib.request.urlopen(req, timeout=120) as r:
        return json.load(r)["data"]["image_token"]


def decimate(task_id: str, key: str, tri_target: int, texture_size: int, out: Path) -> Path:
    """Runs a generation task's result through convert_model and downloads it."""
    conv = post("task", {
        "type": "convert_model",
        "original_model_task_id": task_id,
        "format": "GLTF",
        "face_limit": tri_target,
        "texture_size": texture_size,
    }, key)["task_id"]
    result = await_task(conv, key, "decimate")
    download(result["model"], out)
    return out


def download(url: str, out: Path) -> Path:
    """urlretrieve would send the blocked default User-Agent here too."""
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    with urllib.request.urlopen(req, timeout=300) as r:
        out.write_bytes(r.read())
    return out
